File: app/src/main/java/io/github/golok56/manajemenuang/viewmodels/TransactionDetailViewModel.java
package io.github.golok56.manajemenuang.viewmodels;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import io.github.golok56.manajemenuang.database.AppDatabase;
import io.github.golok56.manajemenuang.models.Transaction;
import io.github.golok56.manajemenuang.utility.Formatter;
import io.github.golok56.manajemenuang.utility.PreferenceManager;
import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * @author Satria Adi Putra
 */

public class TransactionDetailViewModel extends AndroidViewModel {

    private AppDatabase mDatabase;
    private PreferenceManager mPref;

    private MutableLiveData<Transaction> mTransaction = new MutableLiveData<>();

    public TransactionDetailViewModel(@NonNull Application application) {
        super(application);
        mDatabase = AppDatabase.getInstance(application);
        mPref = PreferenceManager.getInstance(application);
    }

    public MutableLiveData<Transaction> getTransaction() {
        return mTransaction;
    }

    public long getId() {
        return mTransaction.getValue().getId();
    }

    public void setId(long id) {
        mDatabase.getTransactionTable().getTransaction(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(transaction -> mTransaction.setValue(transaction));
    }

    public String getAmount() {
        return Formatter.INSTANCE.formatCurrency(mTransaction.getValue().getAmount());
    }

    public String getType() {
        switch (mTransaction.getValue().getType()) {
            case Transaction.EXPENSE_TYPE:
                return "Pengeluaran";
            case Transaction.INCOME_TYPE:
                return "Pemasukan";
        }
        return "";
    }

    public String getDate() {
        return mTransaction.getValue().getDate();
    }

    public void delete() {
        Completable.fromAction(() -> mDatabase.getTransactionTable().delete(mTransaction.getValue()))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }

    public void revert() {
        mPref.revert(mTransaction.getValue());
    }

}

File: app/src/main/java/io/github/golok56/manajemenuang/viewmodels/TransactionDetailViewModel.kt
package io.github.golok56.manajemenuang.viewmodels

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.MutableLiveData

import io.github.golok56.manajemenuang.database.AppDatabase
import io.github.golok56.manajemenuang.models.Transaction
import io.github.golok56.manajemenuang.utility.Formatter
import io.github.golok56.manajemenuang.utility.PreferenceManager
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * @author Satria Adi Putra
 */

class TransactionDetailViewModel(application: Application) : AndroidViewModel(application) {

    private val mDatabase: AppDatabase = AppDatabase.getInstance(application)
    private val mPref: PreferenceManager = PreferenceManager.getInstance(application)

    val transaction = MutableLiveData<Transaction>()

    var id: Long
        get() {
            return transaction.value!!.id
        }
        set(id) {
            mDatabase.transactionTable.getTransaction(id)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe { transaction -> this.transaction.setValue(transaction) }
        }

    val amount: String
        get() = Formatter.formatCurrency(transaction.value!!.amount)

    val type: String
        get() {
            when (transaction.value!!.type) {
                Transaction.EXPENSE_TYPE -> return "Pengeluaran"
                Transaction.INCOME_TYPE -> return "Pemasukan"
            }
            return ""
        }

    val date: String?
        get() = transaction.value!!.date

    fun delete() {
        Completable.fromAction { mDatabase.transactionTable.delete(transaction.value!!) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe()
    }

    fun revert() {
        mPref.revert(transaction.value!!)
    }

}
